third_party/accessibility/ax/platform/ax_platform_node_textprovider_win.cc

```diff
@@ -0,0 +1,354 @@
+
+  // anchor_offset corresponds to the selection start index
+  // and focus_offset is where the selection ends.
+  auto start_offset = unignored_selection.anchor_offset;
+  auto end_offset = unignored_selection.focus_offset;
+
+  // If there's no selected object, return success and don't fill the SAFEARRAY.
+  if (!anchor_object || !focus_object)
+    return S_OK;
+
+  AXNodePosition::AXPositionInstance start =
+      anchor_object->GetDelegate()->CreateTextPositionAt(start_offset);
+  AXNodePosition::AXPositionInstance end =
+      focus_object->GetDelegate()->CreateTextPositionAt(end_offset);
+
+  BASE_DCHECK(!start->IsNullPosition());
+  BASE_DCHECK(!end->IsNullPosition());
+
+  // Reverse start and end if the selection goes backwards
+  if (*start > *end)
+    std::swap(start, end);
+
+  Microsoft::WRL::ComPtr<ITextRangeProvider> text_range_provider =
+      AXPlatformNodeTextRangeProviderWin::CreateTextRangeProvider(
+          std::move(start), std::move(end));
+  if (&text_range_provider == nullptr)
+    return E_OUTOFMEMORY;
+
+  // Since we don't support disjoint text ranges, the SAFEARRAY returned
+  // will always have one element
+  base::win::ScopedSafearray selections_to_return(
+      SafeArrayCreateVector(VT_UNKNOWN /* element type */, 0 /* lower bound */,
+                            1 /* number of elements */));
+
+  if (!selections_to_return.Get())
+    return E_OUTOFMEMORY;
+
+  LONG index = 0;
+  HRESULT hr = SafeArrayPutElement(selections_to_return.Get(), &index,
+                                   text_range_provider.Get());
+  BASE_DCHECK(SUCCEEDED(hr));
+
+  // Since BASE_DCHECK only happens in debug builds, return immediately to
+  // ensure that we're not leaking the SAFEARRAY on release builds
+  if (FAILED(hr))
+    return E_FAIL;
+
+  *selection = selections_to_return.Release();
+
+  return S_OK;
+}
+
+HRESULT AXPlatformNodeTextProviderWin::GetVisibleRanges(
+    SAFEARRAY** visible_ranges) {
+  UIA_VALIDATE_TEXTPROVIDER_CALL();
+
+  const AXPlatformNodeDelegate* delegate = owner()->GetDelegate();
+
+  // Get the Clipped Frame Bounds of the current node, not from the root,
+  // so if this node is wrapped with overflow styles it will have the
+  // correct bounds
+  const gfx::Rect frame_rect = delegate->GetBoundsRect(
+      AXCoordinateSystem::kFrame, AXClippingBehavior::kClipped);
+
+  const auto start = delegate->CreateTextPositionAt(0);
+  const auto end = start->CreatePositionAtEndOfAnchor();
+  BASE_DCHECK(start->GetAnchor() == end->GetAnchor());
+
+  // SAFEARRAYs are not dynamic, so fill the visible ranges in a vector
+  // and then transfer to an appropriately-sized SAFEARRAY
+  std::vector<Microsoft::WRL::ComPtr<ITextRangeProvider>> ranges;
+
+  auto current_line_start = start->Clone();
+  while (!current_line_start->IsNullPosition() && *current_line_start < *end) {
+    auto current_line_end = current_line_start->CreateNextLineEndPosition(
+        AXBoundaryBehavior::CrossBoundary);
```

Comment:
Follow-up to this conversation: https://github.com/flutter/engine/pull/38284#discussion_r1057955169

> @loic-sharma: Could you explain why AXBoundaryDetection::kDontCheckInitialPosition was removed here?

> @yaakovschectman:
The method signature for our version of CreateNextLineEndPosition takes a parameter of a different type than Chrome's current version of the same. We take a AXBoundaryBehavior, while Chromium takes a struct consisting of AXBoundaryBehavior, AXBoundaryDetection

Why are our method signatures different? Should we update our signatures to match Chromium's?

Reply:
This is not only a superficial signature difference, our implementation is different also. We do not have the `AXMovementOptions` struct that exists in Chromium. Updating it in our existing code may warrant a separate PR.

Reply:
Right. What are the resulting behavior differences while we don't have `AXMovementOptions`?

Reply:
Presumably, there will not be an inherent difference between checking or not checking the initial position when moving by a boundary. I think for our purposes a similar effect can be achieved by the conditions used where the enclosing methods are called.